libfwupdplugin/fu-pefile-firmware.c

```diff
@@ -62,124 +103,49 @@ fu_pefile_firmware_parse(FuFirmware *firmware,
 {
 	gsize bufsz = 0;
 	gsize hdr_offset = offset;
-	guint16 opt_hdrsz = 0x0;
-	guint32 nr_sections = 0x0;
-	guint32 offset_sig = 0x0;
-	guint32 signature = 0x0;
+	gsize strtab_offset;
+	guint32 nr_sections;
 	const guint8 *buf = g_bytes_get_data(fw, &bufsz);
+	g_autoptr(GByteArray) st_coff = NULL;
+	g_autoptr(GByteArray) st_doshdr = NULL;
 
-	/* we already checked the MS-DOS magic, check the signature now */
-	if (!fu_memread_uint32_safe(buf,
-				    bufsz,
-				    hdr_offset + 0x3C,
-				    &offset_sig,
-				    G_LITTLE_ENDIAN,
-				    error))
+	/* parse the DOS header to get the COFF header offset */
+	st_doshdr = fu_struct_pe_dos_header_parse(buf, bufsz, offset, error);
+	if (st_doshdr == NULL)
 		return FALSE;
-	hdr_offset += offset_sig;
-	if (!fu_memread_uint32_safe(buf, bufsz, hdr_offset, &signature, G_LITTLE_ENDIAN, error))
+	hdr_offset += fu_struct_pe_dos_header_get_lfanew(st_doshdr);
+	st_coff = fu_struct_pe_coff_header64_parse(buf, bufsz, hdr_offset, error);
+	if (st_coff == NULL)
 		return FALSE;
-	if (signature != 0x4550) {
-		g_set_error_literal(error,
-				    FWUPD_ERROR,
-				    FWUPD_ERROR_INVALID_FILE,
-				    "invalid signature for file");
-		return FALSE;
-	}
 	hdr_offset += 4;
```

Comment:
I think that offset has become more confusing. We add 4 and `FU_STRUCT_PE_COFF_HEADER64_OFFSET_SIZE_OF_OPTIONAL_HEADER`. But what we want is the offset after the coff header (before optional header)

Reply:
Right, agreed. Fixed.